substrate/frame/asset-conversion/src/lib.rs

```diff
@@ -751,157 +749,279 @@ pub mod pallet {
 		/// respecting `keep_alive`.
 		///
 		/// If successful returns the amount of the `path[0]` taken to provide `path[1]`.
-		pub fn do_swap_tokens_for_exact_tokens(
+		///
+		/// WARNING: This may return an error after a partial storage mutation. It should be used
+		/// only inside a transactional storage context and an Err result must imply a storage
+		/// rollback.
+		pub(crate) fn do_swap_tokens_for_exact_tokens(
 			sender: T::AccountId,
-			path: BoundedVec<T::MultiAssetId, T::MaxSwapPathLength>,
-			amount_out: T::AssetBalance,
-			amount_in_max: Option<T::AssetBalance>,
+			path: Vec<T::MultiAssetId>,
+			amount_out: T::Balance,
+			amount_in_max: Option<T::Balance>,
 			send_to: T::AccountId,
 			keep_alive: bool,
-		) -> Result<T::AssetBalance, DispatchError> {
+		) -> Result<T::Balance, DispatchError> {
 			ensure!(amount_out > Zero::zero(), Error::<T>::ZeroAmount);
 			if let Some(amount_in_max) = amount_in_max {
 				ensure!(amount_in_max > Zero::zero(), Error::<T>::ZeroAmount);
 			}
 
 			Self::validate_swap_path(&path)?;
+			let path = Self::balance_path_from_amount_out(amount_out, path)?;
 
-			let amounts = Self::get_amounts_in(&amount_out, &path)?;
-			let amount_in =
-				*amounts.first().defensive_ok_or("get_amounts_in() returned an empty result")?;
-
+			let amount_in = path.first().map(|(_, a)| *a).ok_or(Error::<T>::InvalidPath)?;
```

Comment:
Does the variable `a` stand for the "asset" or "amount"?

Reply:
I mean, we should avoid having those names like 'a,b,c,p,r...'

Reply:
`amount_in` should tell you it is an amount